api/logs/v1/rules_labels_enforcer.go

```diff
@@ -0,0 +1,410 @@
+			return nil
+		}
+
+		var (
+			matchers    = extractMatchers(res.Request, keys)
+			contentType = res.Header.Get("Content-Type")
+		)
+
+		matcherStr := fmt.Sprintf("%s", matchers)
+		level.Debug(logger).Log("msg", "filtering using matchers", "tenant", tenant, "matchers", matcherStr)
+
+		body, err := io.ReadAll(res.Body)
+		if err != nil {
+			level.Error(logger).Log("msg", err)
+			return err
+		}
+		res.Body.Close()
+
+		b, err := filterRules(body, contentType, matchers)
+		if err != nil {
+			level.Error(logger).Log("msg", err)
+			return err
+		}
+
+		res.Body = io.NopCloser(bytes.NewReader(b))
+		res.ContentLength = int64(len(b))
+
+		return nil
+	}
+}
+
+func extractMatchers(r *http.Request, l []string) map[string]string {
+	queryParams := r.URL.Query()
+	matchers := map[string]string{}
+	for _, name := range l {
+		value := queryParams.Get(name)
+		if value != "" {
+			matchers[name] = value
+		}
+	}
+
+	return matchers
+}
+
+func filterRules(body []byte, contentType string, matchers map[string]string) ([]byte, error) {
+	switch contentType {
+	case contentTypeApplicationJSON:
+		var res prometheusRulesResponse
+		err := json.Unmarshal(body, &res)
+		if err != nil {
+			return nil, err
+		}
+
+		return json.Marshal(filterPrometheusResponse(res, matchers))
+
+	case contentTypeApplicationYAML:
+		var res lokiRulesResponse
+		if err := yaml.Unmarshal(body, &res); err != nil {
+			return nil, err
+		}
+
+		return yaml.Marshal(filterLokiRules(res, matchers))
+
+	default:
+		return nil, errUnknownRulesContentType
+	}
+}
+
+func filterPrometheusResponse(res prometheusRulesResponse, matchers map[string]string) prometheusRulesResponse {
+	if len(matchers) == 0 {
+		res.Data = rulesData{}
+		return res
+	}
+
+	if len(res.Data.RuleGroups) > 0 {
+		filtered := filterPrometheusRuleGroups(res.Data.RuleGroups, matchers)
+		res.Data = rulesData{RuleGroups: filtered}
+	}
+
+	if len(res.Data.Alerts) > 0 {
+		filtered := filterPrometheusAlerts(res.Data.Alerts, matchers)
+		res.Data = rulesData{Alerts: filtered}
+	}
+
+	return res
+}
+
+func filterPrometheusRuleGroups(groups []*ruleGroup, matchers map[string]string) []*ruleGroup {
+	var filtered []*ruleGroup
+
+	for _, group := range groups {
+		var filteredRules []rule
+
+	rules:
+		for _, rule := range group.Rules {
+			for key, value := range matchers {
+				if !rule.Labels().Has(key) || rule.Labels().Get(key) != value {
+					continue rules
+				}
+			}
+
+			filteredRules = append(filteredRules, rule)
+		}
+
+		if len(filteredRules) > 0 {
+			group.Rules = filteredRules
+			filtered = append(filtered, group)
+		}
+	}
+
+	return filtered
+}
+
+func filterPrometheusAlerts(alerts []*alert, matchers map[string]string) []*alert {
+	var filtered []*alert
+
+alerts:
+	for _, alert := range alerts {
+		for key, value := range matchers {
+			if !alert.Labels.Has(key) || alert.Labels.Get(key) != value {
+				continue alerts
+			}
+		}
+
+		filtered = append(filtered, alert)
+	}
+
+	return filtered
+}
```

Comment:
Following the suggestion above:
```suggestion
func filterPrometheusAlerts(alerts []*alert, matchers map[string]string) []*alert {
	var filtered []*alert
	for _, alert := range alerts {
		if ruleLabelsMatcheMatchers(alert, matchers) {
			filtered = append(filtered, alert)
		}
	}

	return filtered
}
```